### PR TITLE
Supports more than one ACS location

### PR DESCRIPTION
**Pull request by Shaky212**
see: https://www.pivotaltracker.com/story/show/170045567

### CODE REVIEW

File: src/Surfnet/ServiceProviderDashboard/Domain/Entity/Entity/MetaData.php [hunk 1]
```diff
@@ -79,14 +82,22 @@ public static function fromApiResponse(array $data)
...
         Assert::string($nameEn);
         Assert::string($nameNl);
 
+        if (count(array_unique($acsLocations)) !== count($acsLocations)) {
+            throw new Exception('Double acs locations. Expected unique locations');
+        }
+
+        if (count($acsLocations) > self::MAX_ACS_LOCATIONS) {
+            throw new Exception('Maximum acs locations exceeded. Maximum 10 asc location are supported');
+        }
```
Comment (MKodde) [inline src/Surfnet/ServiceProviderDashboard/Domain/Entity/Entity/MetaData.php:99]: I love you added these input validations! 

1. Must we fail so hard on these conditions? Or would a log message be good enough? Or maybe just ignore the excessive items, or.. One to run by @quartje ?
2. There is a 'lil typo in the second exception a~s~cs

File: src/Surfnet/ServiceProviderDashboard/Domain/ValueObject/Metadata.php [hunk 2]
```diff
@@ -128,6 +128,11 @@ class Metadata
      */
     public $organizationTypeAttribute;
 
+    /**
+     * @var Attribute
+     */
+    public $organizationUnitAttribute;
+
```
Comment (MKodde) [inline src/Surfnet/ServiceProviderDashboard/Domain/ValueObject/Metadata.php:135]: Please scrap this from the commit. It's a patch to work with the older Manage instance in our container

File: src/Surfnet/ServiceProviderDashboard/Infrastructure/DashboardBundle/Form/Entity/SamlEntityType.php [hunk 3]
```diff
@@ -99,14 +100,24 @@ public function buildForm(FormBuilderInterface $builder, array $options)
...
+                            'allow_add' => true,
+                            'allow_delete' => true,
                             'required' => false,
+                            'entry_type' => TextType::class,
+                            'entry_options' => [
+                                'attr' => [
+                                    'data-parsley-redirecturis_set' => 'true',
+                                    'data-parsley-redirecturis_valid' => 'true',
+                                    'data-parsley-urlstrict' => null,
+                                ],
```
Comment (MKodde) [inline src/Surfnet/ServiceProviderDashboard/Infrastructure/DashboardBundle/Form/Entity/SamlEntityType.php:117]: Stricktly speaking these parsley validations are for the redirect urls. In an ideal world we would create a set of validations for the acs locations. A prime example of where DRY is not always the rule to follow.

File: tests/unit/Application/Metadata/JsonGeneratorTest.php [hunk 4]
```diff
@@ -198,6 +198,24 @@ public function test_it_can_build_saml_entity_data_for_new_entities()
...
+                            'AssertionConsumerService:5:Binding' => null,
+                            'AssertionConsumerService:5:Location' => null,
+                            'AssertionConsumerService:6:Binding' => null,
+                            'AssertionConsumerService:6:Location' => null,
+                            'AssertionConsumerService:7:Binding' => null,
+                            'AssertionConsumerService:7:Location' => null,
+                            'AssertionConsumerService:8:Binding' => null,
+                            'AssertionConsumerService:8:Location' => null,
+                            'AssertionConsumerService:9:Binding' => null,
+                            'AssertionConsumerService:9:Location' => null,
```
Comment (MKodde) [inline tests/unit/Application/Metadata/JsonGeneratorTest.php:218]: When adding a new entity. We should not have to add the empty ACS entries.

File: tests/unit/Domain/Entity/Entity/MetaDataTest.php [hunk 5]
```diff
@@ -28,6 +29,125 @@
...
+        $this->assertEquals(
+            null,
+            $metadata['AssertionConsumerService:9:Location']
+        );
+    }
+
+    public function test_it_adds_consecutive_acs_locations()
+    {
+        $metadata = [];
+        AcsLocationHelper::addAcsLocationsToMetaData([null, null, 'https://example1.com', null, 'https://example2.com' ], $metadata);
```
Comment (MKodde) [inline tests/unit/Domain/Entity/Entity/MetaDataTest.php:139]: :ok_hand:

**Review by MKodde — CHANGES_REQUESTED** [review]
# Non code review
I see a lot of great things here @Shaky212 ! 

## Functional test
I have not yet test driven these changes in my browser. But I'm planning to do this tomorrow. I might add some feedback after that testing session. 

## Git history feedback

1. Write message in imperative mood
> Add~ed~ licence header
> Add~ed~ new line at EOF 
2. Aforementioned commits (sh|c)ould have been added to the previous commit (commit --amend && force push). In other words, refrain from adding insignificant commits to your commit history. If the commit is burried deeper in your commit history. Then `commit fixup` can help out! I'd love to show you that in the near future. For now consider squashing the three messages back into your initial commit message. This can be done via an interactve rebase. I'd also like to demonstrate this to you if you are not yet familiar with that little gem :)
3. Your commit description could have been better. The reference to the Pivotal issue is a  good starting point to figure out why you authored this commit. But having that information in the git log is also valuable. Sometimes it feels like you are repeating yourself. But it also helps in writing a good atomic commit.

File: src/Surfnet/ServiceProviderDashboard/Application/Command/Entity/SaveSamlEntityCommand.php [hunk 6]
```diff
@@ -87,15 +87,20 @@ class SaveSamlEntityCommand implements SaveEntityCommandInterface
...
-     *      message = "url.notSecure"
+     * @Assert\All({
+     *      @Assert\NotBlank(),
+     *      @Assert\Url(
+     *          protocols={"https"},
+     *          message = "url.notSecure"
+     *      )
+     * })
+     * @Assert\Count(
+     *     max = 10
```
[the remark this answers is not in the document]
Comment (Shaky212) [reply src/Surfnet/ServiceProviderDashboard/Application/Command/Entity/SaveSamlEntityCommand.php:100]: Message tuned with @quartje has become: 10 ACS locations  or less are allowed

File: src/Surfnet/ServiceProviderDashboard/Application/Metadata/AcsLocationHelper.php [hunk 7]
```diff
@@ -0,0 +1,72 @@
+<?php
+
+/**
+ * Copyright 2017 SURFnet B.V.
```
[the remark this answers is not in the document]
Comment (Shaky212) [reply src/Surfnet/ServiceProviderDashboard/Application/Metadata/AcsLocationHelper.php:4]: Ah, the well known copy/paste ...error action.

File: src/Surfnet/ServiceProviderDashboard/Application/Metadata/AcsLocationHelper.php [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+    /**
+     * Remove empty and null elements from an array.
+     *
+     * @param $array
+     * @return array
+     */
+    public static function cleanArray(&$array)
+    {
+        return array_values(array_filter($array));
+    }
```
[the remark this answers is not in the document]
Comment (Shaky212) [reply src/Surfnet/ServiceProviderDashboard/Application/Metadata/AcsLocationHelper.php:35]: I will turn the method into private access and use it.

File: src/Surfnet/ServiceProviderDashboard/Domain/ValueObject/Metadata.php (hunk 2 above)
Comment (Shaky212) [reply src/Surfnet/ServiceProviderDashboard/Domain/ValueObject/Metadata.php:135]: Ok

File: tests/unit/Application/Metadata/JsonGeneratorTest.php (hunk 4 above)
Comment (Shaky212) [reply tests/unit/Application/Metadata/JsonGeneratorTest.php:218]: Used the $entity->isManageEntity() for new/existing entities. Also added function to add  empty ACS locations up to the max allowed locations.

File: src/Surfnet/ServiceProviderDashboard/Domain/Entity/Entity/MetaData.php (hunk 1 above)
Comment (Shaky212) [reply src/Surfnet/ServiceProviderDashboard/Domain/Entity/Entity/MetaData.php:99]: After consultation with Bart. Manage does not validate whether two of the same ACS locations exist. It is useful to perform that validation client side on our side.

File: src/Surfnet/ServiceProviderDashboard/Application/Metadata/AcsLocationHelper.php [hunk 9]
```diff
@@ -0,0 +1,72 @@
...
+     * imports the POST ACS URLs. Other formats are not supported by manage or
+     * the dashboard.
+     *
+     * @param $acsLocations
+     * @param $metadata
+     * @return void
+     */
+    public static function addAcsLocationsToMetaData($acsLocations, &$metadata)
+    {
+        $ascLocations = array_values(array_filter($acsLocations));
```
[the remark this answers is not in the document]
Comment (Shaky212) [reply src/Surfnet/ServiceProviderDashboard/Application/Metadata/AcsLocationHelper.php:49]: Used the private cleanArray method here.

**Pull request merged**